rust-symcrypt/src/hash.rs

```diff
@@ -62,14 +62,16 @@ impl Drop for Sha256State {
     }
 }
 
-pub fn sha256(data: &[u8], result: &mut [u8; SHA256_RESULT_SIZE]) {
+pub fn sha256(data: &[u8]) -> [u8; SHA256_RESULT_SIZE] {
```

Comment:
Choosing [u8; SHA256_RESULT_SIZE] over vec because size is known. Would there be any drawbacks from this? I understand that vec is more common and easier to work with in Rust.

Reply:
In my opinion, `[u8; SHA256_RESULT_SIZE]` is better than `Vec<u8>` in pretty much every way given that you know the size is always going to be exactly `SHA256_RESULT_SIZE`. The most important difference is that `Vec` always involves a heap allocation, while the fixed sized array can be stored inline. If you want to put store the result on the heap, you can do `Box::new(sha256(data))`, which gives you something similar to but not quite a `Vec`. 

Another option to consider is making the result something like `Sha256Result`, defined like this:

```rust
struct Sha256Result {
    data: [u8; SHA256_SIZE],
}
```

Basically, the hides the fact that the hash is an array of bytes, which may or may not be useful. It depends on whether you want people to manipulate the bytes directly, or if the hash should always be treated semantically as a single unit?

Reply:
I think returning the array would be easier because it gives the user an option if needed to modify/splice easily, it is also what is returned in symcrypt. 

On the other hand the function that I'm working on right now; `SymCryptCGM*` can fail and will need to have a struct to wrap the symcrypt errors/success. If we're going to implement a return type for the `SymCryptGCM*`, does it make sense to have wrappers for the other functions that don't necessarily need it? ( Hmac, Hash )

Reply:
So for functions that can fail, the normal thing to do is return [`Result`](https://doc.rust-lang.org/std/result/). The `Ok` part is just whatever type the function would return on success. For FFI crates like this, the `Err` part usually wraps some enum that lists all the errors that could be returned by SymCrypt. It's good practice to use [thiserror](https://crates.io/crates/thiserror) to `#[derive(Debug)]` on your error type as well because that will make it play better with the rest of Rust's error handling.